### PR TITLE
[DEX-146] extend api POST creation of encounter to include references to assets

### PR DESCRIPTION
**Pull request by naknomum**
## Pull Request Overview

- allows api json payload to create encounters include `.assets` array to be attached
- many-to-many class **EncounterAssets** created to link encounters and assets
- also some bug fixes and completion of DELETE passthrough (corresponding to edm support of same)

---

**Review Notes**
- mostly just building around optional assets hanging off of the houston feather model for encounter.  e.g. encounter.delete() created which properly cleans up these cascading depedencies.
- tweaked exact api call for GET encounter, which now also includes `.assets: [ ]`
- did _some_ modifications of owner/public on encounter -- but _more is needed_ especially after the "spirited"(?) discussion howard, colin, and i had today -- this will be its own PR forthcoming
- ~~**_unresolved_**: one passthrough failure case is not yet handled well; want to discuss my plan and alternative tactics friday. similar scenarios likely for other houston<->edm exchanges as well.~~ (fixed via subsequent commit after feedback)
- no tests on any stuff that _requires edm_ still.  maybe we can discuss friday as well.  perhaps stop-gap solution of a single edm instance to test against is in order until contractor-folks get something more docker-based?

**Database Updates**

new relationship EncounterAssets introduced to join encounter/assets.  e.g. `encounter.assets`, table: `encounter_assets`

**REST API Updates**

- `api/v1/encounters` POST now can accept optional addition of:
```
"assets": [
    {   "guid": "f373305f-18ca-4244-ac33-8a8f648dde9d"   },
    . . .
]
```

- previously `api/v1/encounters/GUID/complete` was to test getting a full json object, but dropped the "/complete" so its now just the standard GET path to an object

**Pull Request Checklist**
- [x] Ensure that the PR is properly formatted
  - Example: All lint checks are passing
- [x] Ensure that the PR is properly rebased
  - Example: The PR is rebased on `develop` (commit: `e7cfeb4c`)
- [x] Ensure that the PR uses a consolidated database migration
  - Example: One database migration is proposed (revision `787b9a5d1ef5` -> `5854c14c634d`)
- [x] Ensure that the PR is properly tested
  - Example: All automated tests are passing
- [x] Ensure that the PR is properly sanitized
  - Example: No sensitive data or large content was added to this PR
- [ ] Ensure that the PR is properly reviewed

### CODE REVIEW

File: app/modules/encounters/resources.py [hunk 1]
```diff
@@ -107,20 +107,50 @@ def post(self, args):
...
+                    pub = False
+                encounter = Encounter(
+                    guid=result_data['id'],
+                    version=result_data.get('version', 2),
+                    owner_guid=owner_guid,
+                    public=pub,
+                )
+                encounter.assets = asset_refs
+                db.session.add(encounter)
+        except Exception as ex:
```
Comment (naknomum) [inline app/modules/encounters/resources.py:142]: _this_ is the point i wish to discuss on friday.  there are a few permutations of failure cases i try to recover from here.  the one i need some closure on is when the **edm _successfully_ creates** its object (the "primary" one), and **houston fails** to create/persist its feather equivalent right after that.
currently i am leaning toward cleaning up by calling back to edm to delete what it just made (see line 151), but there are other ways to look at this -- and probably other permutations i havent considered.

File: app/modules/users/models.py [hunk 2]
```diff
@@ -595,3 +597,12 @@ def delete(self):
         with db.session.begin():
             # TODO: Ensure proper cleanup
             db.session.delete(self)
+
+    @classmethod
+    def get_public_user(cls):
```
Comment (naknomum) [inline app/modules/users/models.py:602]: this snuck in based on some stuff colin and i were talking about; really has no relevance until the PR related to owner/submitter which will happen shortly....

File: app/modules/encounters/resources.py [hunk 3]
```diff
@@ -204,63 +254,23 @@ def delete(self, encounter):
...
+            )
 
         # if we get here, edm has deleted the encounter, now houston feather
-        context = api.commit_or_abort(
-            db.session, default_error_message='Failed to delete the Encounter.'
-        )
         # TODO handle failure of feather deletion (when edm successful!)  out-of-sync == bad
-        with context:
-            db.session.delete(encounter)
+        encounter.delete()
```
Comment (naknomum) [inline app/modules/encounters/resources.py:270]: passthrough-delete has some potential to get out-of-sync as well (in addition to create/POST, i mean)... but is less volatile.  we also will need some integrity check to find/repair these sync woes that inevitably will pop up.  am thinking of writing 1 or 2 invoke-app-tasks for this purpose soon.  (but seems a little lower priority now -- maybe just a jira ticket?)

File: app/modules/encounters/resources.py (hunk 1 above)
Comment (naknomum) [reply app/modules/encounters/resources.py:142]: discussion with @colinwkingen concluded to go with edm-object delete as recovery behavior here.  basically, cleanup/undo what has happened and then fail on the api.  going to add that asap.

**Review by colinwkingen — APPROVED** [review]
LGTM

**Comment by bluemellophone** [comment]
The CodeCov coverage regression will be addressed in a future PR with branch `enc_tests`

**Pull request merged**